bmq-sdk/src/main/java/com/bloomberg/bmq/impl/infr/io/ByteBufferOutputStream.java

```diff
@@ -104,26 +121,42 @@ public void writeByte(int v) throws IOException {
         currentBuffer.put((byte) v);
     }
 
+    @Override
     public void writeBytes(String s) throws IOException {
         throw new UnsupportedOperationException();
     }
 
-    public void writeBytes(ByteBuffer b) throws IOException {
+    private boolean bufferIsFresh(ByteBuffer b) {
+        // a buffer that has never been put() to nor flipped
```

Comment:
this is one thing that really bugs me about nio.ByteBuffer, there is no way to determine if a buffer that has no data vs one that has data but has already been flipped for reading.  `isClear()` is probably a better name for this method ...

but in the long term, I think the better thing to do is replace every usage of nio.ByteBuffer with netty.ByteBuf in the entire SDK, which would
1. allow to eliminate that initial copy from NettyTcpConnection
2. allow to delete ByteBufferOutputStream and ByteBufferInputStream entirely since netty.ByteBufOutputStream can be used instead

Reply:
> this is one thing that really bugs me about nio.ByteBuffer, there is no way to determine if a buffer that has no data vs one that has data but has already been flipped for reading. `isClear()` is probably a better name for this method ...

Agree, but maybe we do not need to determine this and just take what is between 0 and limit regardless of current position? 

> but in the long term, I think the better thing to do is replace every usage of nio.ByteBuffer with netty.ByteBuf in the entire SDK, which would
> 
> 1. allow to eliminate that initial copy from NettyTcpConnection
> 2. allow to delete ByteBufferOutputStream and ByteBufferInputStream entirely since netty.ByteBufOutputStream can be used instead

I would agree but AFAIR the limited use of netty and its components was intentional. The reason is that in some day SDK may switch to another network library and in case of wide use of its components in the code e.g. `netty.ByteBuf` instead of `nio.ByteBuffer` etc. it would require a lot of work to do. And having `netty.ByteBuf` in API class (`Queue`) will make it even more difficult.

I suggest to have a quick call tomorrow to discuss this